File: mobile/ios/CompanionRobot/BLE/BcpCodec.swift
import Foundation

/// BCP (Bundle Command Protocol) codec — Swift implementation.
///
/// Wire format (little-endian):
///   Header:  Magic(1B) + Version(1B) + TotalLen(2B LE) + SeqNo(2B LE) + CmdCount(1B) + Reserved(1B)
///   Commands: CmdID(2B LE) + PayloadLen(1B) + Payload
///   CRC: CRC-16/CCITT over bytes 0..TotalLen-2

struct BcpCodec {
    static let magic: UInt8 = 0xCB
    static let version: UInt8 = 0x01
    static let headerLen = 8
    static let crcLen = 2
    static let maxFrameLen = 1024
    static let maxCommands = 32
    static let maxPayloadLen = 255

    // Command IDs
    static let cmdMove: UInt16     = 0x0101
    static let cmdStop: UInt16     = 0x0105
    static let cmdLedSet: UInt16   = 0x0201
    static let cmdLedOff: UInt16   = 0x0203
    static let cmdFaceExpr: UInt16 = 0x0204
    static let cmdSpeak: UInt16    = 0x0206
    static let cmdServoSet: UInt16 = 0x0107
    static let cmdHeartbeat: UInt16 = 0x0001

    // ── CRC-16/CCITT ───────────────────────────────────────────

    private static let crcTable: [UInt16] = {
        var table = [UInt16](repeating: 0, count: 256)
        for i in 0..<256 {
            var crc = UInt16(i) << 8
            for _ in 0..<8 {
                crc = (crc & 0x8000) != 0 ? (crc << 1) ^ 0x1021 : crc << 1
            }
            table[i] = crc
        }
        return table
    }()

    static func crc16(_ data: Data) -> UInt16 {
        var crc: UInt16 = 0xFFFF
        for byte in data {
            let idx = Int((UInt16(crc >> 8) ^ UInt16(byte)) & 0xFF)
            crc = (crc << 8) ^ crcTable[idx]
        }
        return crc
    }

    // ── Types ──────────────────────────────────────────────────

    struct BcpCommand {
        let cmdId: UInt16
        let payload: Data

        var payloadLen: UInt8 { UInt8(payload.count) }
        var wireLen: Int { 3 + payload.count }
    }

    struct BcpFrame {
        let seqNo: UInt16
        let commands: [BcpCommand]

        var cmdCount: Int { commands.count }
        var totalLen: Int {
            headerLen + commands.reduce(0) { $0 + $1.wireLen } + crcLen
        }
    }

    // ── Encode ─────────────────────────────────────────────────

    static func encode(frame: BcpFrame) -> Data {
        var data = Data(capacity: frame.totalLen)

        // Header
        data.append(magic)
        data.append(version)
        data.append(contentsOf: writeU16LE(UInt16(frame.totalLen)))
        data.append(contentsOf: writeU16LE(frame.seqNo))
        data.append(UInt8(frame.cmdCount))
        data.append(0) // reserved

        // Commands
        for cmd in frame.commands {
            data.append(contentsOf: writeU16LE(cmd.cmdId))
            data.append(cmd.payloadLen)
            data.append(cmd.payload)
        }

        // CRC over everything except CRC itself
        let crc = crc16(data)
        data.append(contentsOf: writeU16LE(crc))

        return data
    }

    // ── Helpers ────────────────────────────────────────────────

    private static func writeU16LE(_ value: UInt16) -> [UInt8] {
        [UInt8(value & 0xFF), UInt8((value >> 8) & 0xFF)]
    }
}

File: mobile/android/app/src/main/java/com/companion/robot/ble/BcpCodec.kt
package com.companion.robot.ble

/**
 * BCP (Bundle Command Protocol) codec — Kotlin implementation.
 *
 * Wire format (little-endian):
 *   Header:  Magic(1B) + Version(1B) + TotalLen(2B LE) + SeqNo(2B LE) + CmdCount(1B) + Reserved(1B)
 *   Commands: CmdID(2B LE) + PayloadLen(1B) + Payload(N bytes)
 *   CRC: CRC-16/CCITT over bytes 0..TotalLen-2
 */

object BcpCodec {
    // Protocol constants
    const val MAGIC: Byte = 0xCB.toByte()
    const val VERSION: Byte = 0x01
    const val HEADER_LEN = 8
    const val CRC_LEN = 2
    const val MAX_FRAME_LEN = 1024
    const val MAX_COMMANDS = 32
    const val MAX_PAYLOAD_LEN = 255

    // Command IDs
    const val CMD_HEARTBEAT: Int = 0x0001

    // ── CRC-16/CCITT lookup table ──────────────────────────────

    private val crcTable: Array<Short> by lazy {
        Array(256) { i ->
            var crc = (i shl 8).toShort()
            repeat(8) {
                crc = if ((crc.toInt() and 0x8000) != 0) {
                    ((crc.toInt() shl 1) xor 0x1021).toShort()
                } else {
                    (crc.toInt() shl 1).toShort()
                }
            }
            crc
        }
    }

    fun crc16(data: ByteArray, offset: Int = 0, len: Int = data.size): Short {
        var crc: Short = 0xFFFF.toShort()
        for (i in offset until offset + len) {
            val idx = ((crc.toInt() shr 8) xor (data[i].toInt() and 0xFF)) and 0xFF
            crc = ((crc.toInt() shl 8) xor crcTable[idx].toInt()).toShort()
        }
        return crc
    }

    // ── Encode ─────────────────────────────────────────────────

    data class BcpCommand(val cmdId: Int, val payload: ByteArray) {
        val payloadLen: Int get() = payload.size
        val wireLen: Int get() = 3 + payload.size // cmdId(2) + payloadLen(1) + payload

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is BcpCommand) return false
            return cmdId == other.cmdId && payload.contentEquals(other.payload)
        }

        override fun hashCode(): Int = cmdId * 31 + payload.contentHashCode()
    }

    data class BcpFrame(
        val seqNo: Int,
        val commands: List<BcpCommand>
    ) {
        val cmdCount: Int get() = commands.size
        val totalLen: Int get() = HEADER_LEN + commands.sumOf { it.wireLen } + CRC_LEN

        companion object {
            fun create(seqNo: Int, vararg commands: BcpCommand): BcpFrame {
                return BcpFrame(seqNo, commands.toList())
            }
        }
    }

    fun encode(frame: BcpFrame, buf: ByteArray): Int {
        var pos = 0

        // Header
        buf[pos++] = MAGIC
        buf[pos++] = VERSION
        writeU16LE(buf, pos, frame.totalLen); pos += 2
        writeU16LE(buf, pos, frame.seqNo); pos += 2
        buf[pos++] = frame.cmdCount.toByte()
        buf[pos++] = 0 // reserved

        // Commands
        for (cmd in frame.commands) {
            writeU16LE(buf, pos, cmd.cmdId); pos += 2
            buf[pos++] = cmd.payloadLen.toByte()
            cmd.payload.copyInto(buf, pos)
            pos += cmd.payload.size
        }

        // CRC over everything except CRC field itself
        val crc = crc16(buf, 0, pos)
        writeU16LE(buf, pos, crc.toInt()); pos += 2

        return pos
    }

    // ── Helpers ────────────────────────────────────────────────

    private fun writeU16LE(buf: ByteArray, offset: Int, value: Int) {
        buf[offset] = (value and 0xFF).toByte()
        buf[offset + 1] = ((value shr 8) and 0xFF).toByte()
    }

    fun readU16LE(buf: ByteArray, offset: Int): Int =
        (buf[offset].toInt() and 0xFF) or ((buf[offset + 1].toInt() and 0xFF) shl 8)
}
